### ISSUE TITLE
String.toInt returns a Float in certain circumstances

### ISSUE THREAD

**Issue by ChrisWellsWood** [elm-lang/core#831]
I was doing some validation on input, and noticed an odd situation when parsing strings:

```Elm
> String.toInt "1"
Ok 1 : Result.Result String Int
> String.toInt "dfsf"
Err "could not convert string 'dfsf' to an Int" : Result.Result String Int
> String.toInt "-"
Ok NaN : Result.Result String Int
> String.toInt "+"
Ok NaN : Result.Result String Int
```
But the type of `NaN` is `Float`, not `Int`. I've not found any other circumstances where this occurs yet. I have a work around where I catch it manually, but I have to use `toFloat` on it first in order to use `isNaN`.

I'm using Elm 0.18, on FireFox and Elm repl. Thanks!

**Comment by jvoigtlaender** [comment]
There do exist other operations in current Elm that produce `NaN` despite being typed `Int`. See https://github.com/elm-lang/core/issues/590, covered in https://github.com/elm-lang/core/issues/721.

Also, it would be pretty important that you not only specify which version of Elm you use, but also which version of the `core` library you use. Because the implementation of `String.toInt` was recently changed, see https://github.com/elm-lang/core/releases/tag/5.1.0.

**Comment by mgold** [comment]
This issue reproduces in 0.18.0/5.1.1.

**Comment by ChrisWellsWood** [comment]
Yeah, I'm using Elm 0.18 and elm-core 5.1.1.

**Comment by evancz** [comment]
Fixed by OP in #834
